src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -0,0 +1,146 @@
+  _card_shift = CardTable::card_shift;
+
+  _byte_map = _card_table->byte_for_index(0);
+
+  _whole_heap_base = _card_table->addr_for(_byte_map);
+  _whole_heap_end = _whole_heap_base + total_card_count * CardTable::card_size;
+
+  _byte_map_base = _byte_map - (uintptr_t(_whole_heap_base) >> _card_shift);
+
+  _overreach_map = (uint8_t *) malloc(total_card_count);
+  _overreach_map_base = (_overreach_map -
+                         (uintptr_t(_whole_heap_base) >> _card_shift));
+
+  assert(total_card_count % ShenandoahCardCluster<ShenandoahDirectCardMarkRememberedSet>::CardsPerCluster == 0, "Invalid card count.");
+  assert(total_card_count > 0, "Card count cannot be zero.");
+  // assert(_overreach_cards != NULL);
+}
+
+ShenandoahDirectCardMarkRememberedSet::~ShenandoahDirectCardMarkRememberedSet() {
+  free(_overreach_map);
+}
+
+void ShenandoahDirectCardMarkRememberedSet::initialize_overreach(uint32_t first_cluster, uint32_t count) {
+
+  // We can make this run faster in the future by explicitly
+  // unrolling the loop and doing wide writes if the compiler
+  // doesn't do this for us.
+  uint32_t first_card_index = first_cluster * ShenandoahCardCluster<ShenandoahDirectCardMarkRememberedSet>::CardsPerCluster;
+  uint8_t *omp = &_overreach_map[first_card_index];
+  uint8_t *endp = omp + count * ShenandoahCardCluster<ShenandoahDirectCardMarkRememberedSet>::CardsPerCluster;
+  while (omp < endp)
+    *omp++ = CardTable::clean_card_val();
+}
+
+void ShenandoahDirectCardMarkRememberedSet::merge_overreach(uint32_t first_cluster, uint32_t count) {
+
+  // We can make this run faster in the future by explicitly unrolling the loop and doing wide writes if the compiler
+  // doesn't do this for us.
+  uint32_t first_card_index = first_cluster * ShenandoahCardCluster<ShenandoahDirectCardMarkRememberedSet>::CardsPerCluster;
+  uint8_t *bmp = &_byte_map[first_card_index];
+  uint8_t *endp = bmp + count * ShenandoahCardCluster<ShenandoahDirectCardMarkRememberedSet>::CardsPerCluster;
+  uint8_t *omp = &_overreach_map[first_card_index];
+
+  // dirty_card is 0, clean card is 0xff; if either *bmp or *omp is dirty, we need to mark it as dirty
+  while (bmp < endp)
+    *bmp++ &= *omp++;
+}
+
+
+// Implementation is not correct.  ShenandoahBufferWithSATBRememberedSet is a placeholder for future planned improvements.
+ShenandoahBufferWithSATBRememberedSet::ShenandoahBufferWithSATBRememberedSet(size_t card_count)
+{
+  _heap = ShenandoahHeap::heap();
+
+  _card_count = card_count;
+  _cluster_count = _card_count /
+      ShenandoahCardCluster<ShenandoahBufferWithSATBRememberedSet>::CardsPerCluster;
+  _card_shift = CardTable::card_shift;
+
+  _whole_heap_base = _heap->base();
+  _whole_heap_end = _whole_heap_base + _card_count *
+      ShenandoahCardCluster<ShenandoahBufferWithSATBRememberedSet>::CardsPerCluster;
+}
+
+// Implementation is not correct.  ShenandoahBufferWithSATBRememberedSet is a placeholder for future planned improvements.
```

Comment:
This should go into its own .cpp file I think. Actually, I am not sure that we should add empty&unused code at all. Maybe add it when you also fill the implementation and use?